File: 01. Arrays/04. Maximum Subarray Circular/main.py
def kadane(nums):
    max_sum = nums[0]
    current_sum = nums[0]

    for i in range(1, len(nums)):
        current_sum = max(nums[i], current_sum + nums[i])
        max_sum = max(max_sum, current_sum)

    return max_sum

def max_circular_subarray_sum(nums):
    max_kadane = kadane(nums)
    total_sum = sum(nums)
    min_sum = nums[0]
    current_min = nums[0]

    for i in range(1, len(nums)):
        current_min = min(nums[i], current_min + nums[i])
        min_sum = min(min_sum, current_min)

    return max_kadane if total_sum == min_sum else max(max_kadane, total_sum - min_sum)

def main():
    nums = [5, -3, 5]
    print("Maximum Circular Subarray Sum:", max_circular_subarray_sum(nums))

main()

File: 01. Arrays/04. Maximum Subarray Circular/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int kadane(vector<int>& nums) {
    int maxSum = nums[0], currentSum = nums[0];
    for (size_t i = 1; i < nums.size(); i++) {
        currentSum = max(nums[i], currentSum + nums[i]);
        maxSum = max(maxSum, currentSum);
    }
    return maxSum;
}

int maxCircularSubarraySum(vector<int>& nums) {
    int maxKadane = kadane(nums);
    int totalSum = 0, minSum = nums[0], currentMin = nums[0];

    for (int i = 0; i < nums.size(); i++) {
        totalSum += nums[i];
        if (i > 0) {
            currentMin = min(nums[i], currentMin + nums[i]);
            minSum = min(minSum, currentMin);
        }
    }
    
    if (totalSum == minSum) return maxKadane;
    return max(maxKadane, totalSum - minSum);
}

int main() {
    vector<int> nums = {5, -3, 5};
    cout << "Maximum Circular Subarray Sum: " << maxCircularSubarraySum(nums) << endl;
    return 0;
}
